File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int v, e;
    static List<int[]>[] graph;
    public static class Node implements Comparable<Node> {
        int distance;
        int region;

        public Node(int distance, int region) {
            this.distance = distance;
            this.region = region;
        }

        @Override
        public int compareTo(Node o) {
            return this.distance - o.distance;
        }
    }

    public static int[] dijkstra(int start) {
        int[] distance = new int[v+1];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[start] = 0;
        PriorityQueue<Node> heapQ = new PriorityQueue<>();
        heapQ.add(new Node(0, start));

        while (!heapQ.isEmpty()) {
            Node node = heapQ.poll();

            if (distance[node.region] < node.distance) {
                continue;
            }

            for (int[] dis : graph[node.region]) {
                int cost = node.distance + dis[1];

                if (distance[dis[0]] > cost) {
                    distance[dis[0]] = cost;
                    heapQ.add(new Node(cost, dis[0]));
                }
            }
        }

        return distance;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        v = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());

        graph = new List[v+1];
        for (int i = 0; i < v+1; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            graph[a].add(new int[] {b, c});
            graph[b].add(new int[] {a, c});
        }

        st = new StringTokenizer(br.readLine());
        int v1 = Integer.parseInt(st.nextToken());
        int v2 = Integer.parseInt(st.nextToken());

        int[] totalStart = dijkstra(1);
        int[] v1Start = dijkstra(v1);
        int[] v2Start = dijkstra(v2);

        int v1F = totalStart[v1] + v1Start[v2] + v2Start[v];
        int v2F = totalStart[v2] + v2Start[v1] + v1Start[v];

        int ans = Math.min(v1F, v2F);
        if (ans >= (int) 1e9 || ans < 0) {
            ans = -1;
        }

        System.out.println(ans);
    }
}

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.py
import heapq
import sys
input = sys.stdin.readline

def dijkstra(start):
    distance = [1e9] * (n +1)
    q = []
    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q:
        dist, now = heapq.heappop(q)

        if distance[now] < dist:
            continue

        for i in graph[now]:
            cost = dist + i[1]

            if distance[i[0]] > cost:
                distance[i[0]] = cost
                heapq.heappush(q, (cost, i[0]))

    return distance

n, e = map(int, input().split())

graph = [[] for _ in range (n+1)]

for i in range(e):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))
    graph[b].append((a, c))

v1, v2 = map(int, input().split())

first_start = dijkstra(1)
v1_start = dijkstra(v1)
v2_start = dijkstra(v2)

# 최초에서 v1까지 - v1에서 v2까지 - v2에서 end 까지 거리
v1_f = first_start[v1] + v1_start[v2] + v2_start[n]
# 최초에서 v2까지 - v2에서 v1까지 - v1에서 end 까지 거리
v2_f = first_start[v2] + v2_start[v1] + v1_start[n]

ans = min(v1_f, v2_f)

if ans >= 1e9:
    ans = -1

print(ans)